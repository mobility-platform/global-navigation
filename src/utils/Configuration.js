import { createContext, h } from "preact";
import { useContext, useEffect, useState } from "preact/hooks";

const ConfigurationContext = createContext();

/**
 * @typedef ConfigurationLink
 * @property {string} title The link title (e.g.: MyInspection)
 * @property {string} icon The link icon. Can be an URL to the icon or a SVG element.
 * @property {string} href The link target.
 * @property {boolean} [requiresAuthentification=false] Whether or not this link requires the user to be authenticated to be displayed.
 */

/**
 * @typedef ConfigurationSection
 * @property {string} title The section title (e.g.: Global)
 * @property {ConfigurationLink[]} links The sections links
 * @property {boolean} [requiresAuthentification=false] Whether or not this section requires the user to be authenticated to be displayed.
 */

/**
 * @typedef Configuration
 * @property {string} themeApiUrl The theme API URL (e.g.: https://myapi.services.michelin.com/themes/current). This will be used to fetch the color theme.
 * @property {string} userApiUrl The user profile API URL (e.g.: https://myapi.services.michelin.com/users). This will be used to fetch user profile.
 * @property {string} profileUrl The user profile URL (e.g.: https://mymobility.services.michelin.com/users/me). This will be used when the user click on 'See my profile'.
 * @property {string} notificationsApiUrl The notifications API URL (e.g.: https://mobility-platform-notifications.azurewebsites.net/v1/notifications). Thiw will be used to fetch user notifications and mark them as read.
 * @property {ConfigurationSection[]} sections The link sections to display (e.g.: Global, Apps, etc.).
 */

/***
 * Fetch the Global Navigation configuration located at the given `url` (e.g.: https://unpkg.com/@mobility-platform/global-navigation-configuration@latest/production.config.json) and
 * expose it to child component with the `useConfiguration` hook.
 */
export const ConfigurationProvider = ({ url, children }) => {
  console.log("ConfigurationProvider");
  const configuration = useFetchConfiguration(url);
  return (
    <ConfigurationContext.Provider value={configuration}>{children}</ConfigurationContext.Provider>
  );
};

/**
 * Access the global navigation configuration.
 * @type {() => Configuration}
 */
export const useConfiguration = () => useContext(ConfigurationContext);

export const useFetchConfiguration = (url) => {
  const [state, setState] = useState();
  useEffect(() => {
    fetch(url)
      .then((response) => response.json())
      .then(setState);
  }, [url]);
  return state;
};
