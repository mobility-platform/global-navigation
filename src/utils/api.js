export const fetchTheme = async ({ getToken, apiUrl }) => {
  const token = await getToken();

  const endPoint = `${apiUrl}/themes/current`;

  let response = await fetch(endPoint, {
    headers: {
      Authorization: "Bearer " + token
    }
  });
  let data = await response.json();

  return data;
};

export const fetchApplications = async ({ getToken, apiUrl }) => {
  const token = await getToken();

  const endPoint = `${apiUrl}/applications`;

  let response = await fetch(endPoint, {
    headers: {
      Authorization: "Bearer " + token
    }
  });
  let data = await response.json();

  return data;
};
