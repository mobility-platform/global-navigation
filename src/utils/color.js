function getLuminanceFromRgbValue(value) {
  const normalizedValue = value / 255;
  return 0.03928 > normalizedValue
    ? normalizedValue / 12.92
    : Math.pow((normalizedValue + 0.055) / 1.055, 2.4);
}

function getLuminanceFromRgbColor(r, g, b) {
  return (
    0.2126 * getLuminanceFromRgbValue(r) +
    0.7152 * getLuminanceFromRgbValue(g) +
    0.0722 * getLuminanceFromRgbValue(b)
  );
}

function getLuminanceFromHex(hex) {
  const [, r, g, b] = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return getLuminanceFromRgbColor(parseInt(r, 16), parseInt(g, 16), parseInt(b, 16));
}

export function isTextLegibleOverBackground(textColor, backgroundColor) {
  const textLightness = getLuminanceFromHex(textColor),
    backgroundLightness = getLuminanceFromHex(backgroundColor);
  const contrast =
    (Math.max(textLightness, backgroundLightness) + 0.05) /
    (Math.min(textLightness, backgroundLightness) + 0.05);
  return contrast >= 4.5;
}

export function isDark(background) {
  return isTextLegibleOverBackground("#ffffff", background);
}

export function isLight(background) {
  return !isDark(background);
}

export function getContrastColor(background) {
  return isDark(background) ? "#ffffff" : "#333333";
}
