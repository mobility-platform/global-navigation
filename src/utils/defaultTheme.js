const defaultTheme = {
  color: "#eef5ff",
  primary: "#27509b",
  background: "#FFFFFF"
};
export default defaultTheme;
