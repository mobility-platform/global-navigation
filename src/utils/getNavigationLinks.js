const getNavigationLinks = baseUrl => {
  if (!baseUrl) {
    return;
  }
  return {
    globalLinks: [
      {
        href: `${baseUrl}`,
        label: "My Home",
        icon: `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="17"
          fill="none"
          viewBox="0 0 16 17"
        >
          <defs />
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M1 6.3L8 1l7 5.3v8.2c0 .8-.7 1.5-1.6 1.5H2.6c-.9 0-1.6-.7-1.6-1.5V6.2z"
            clip-rule="evenodd"
          />
        </svg>
        `
      },
      {
        href: `${baseUrl}/groups`,
        label: "My Organization",
        icon: `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          aria-hidden="true"
          viewBox="0 0 448 512"
        >
          <defs />
          <path d="M352 320a95.6 95.6 0 00-59.8 20.9l-102.5-64a96.6 96.6 0 000-41.7L292.2 171a96 96 0 10-34-54.3l-102.4 64a96 96 0 100 150.2l102.5 64.2A96.3 96.3 0 00256 416a96 96 0 1096-96z" />
        </svg>
        `
      }
    ],
    profileLink: {
      href: `${baseUrl}/user/me`,
      label: "View my profile"
    }
  };
};

export default getNavigationLinks;
