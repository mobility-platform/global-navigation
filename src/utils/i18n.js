import { createContext } from "preact";
import { useCallback, useContext, useMemo } from "preact/hooks";

const translations = {
  fr: {
    MyShowcase: "MyShowcase",
    MyShop: "MyShop",
    Global: "General",
    Apps: "Applications",
    Others: "Autres",
    "View my profile": "Voir mon profil",
    "Open the menu": "Ouvrir le menu",
    "Close the menu": "Fermer le menu",
    "You currently have no notifications": "Vous n'avez pas de notifications",
    Login: "Se connecter"
  }
};

/** @type {() => string} */
const useDefaultLanguage = () => {
  return useMemo(() => navigator.languages?.[0] || navigator.language, []);
};

/** @type {(language: string) => string} */
const normalizeLanguage = language => {
  return language.substr(0, 2);
};

/** @type {import("preact").Context<string>} */
const TranslationContext = createContext("en");

/** @type {() => string} */
export const useLanguage = () => {
  const contextLanguage = useContext(TranslationContext);
  const defaultLanguage = useDefaultLanguage();
  const language = useMemo(() => normalizeLanguage(contextLanguage || defaultLanguage), [
    contextLanguage,
    defaultLanguage
  ]);
  return language;
};

export const TranslationProvider = TranslationContext.Provider;

/** @type {(key: string, language: string) => string} */
export const t = (key, language) => translations?.[language]?.[key] || key;

/** @type {() => (key: string) => string} */
export const useTranslation = () => {
  const contextLanguage = useContext(TranslationContext);
  const defaultLanguage = useDefaultLanguage();
  const language = useMemo(() => normalizeLanguage(contextLanguage || defaultLanguage), [
    contextLanguage,
    defaultLanguage
  ]);
  return useCallback(key => t(key, language), [language]);
};
