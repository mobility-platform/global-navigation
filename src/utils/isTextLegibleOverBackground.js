function getLFromRgbValue(value) {
  var normalizedValue = value / 255;
  return 0.03928 > normalizedValue
    ? normalizedValue / 12.92
    : Math.pow((normalizedValue + 0.055) / 1.055, 2.4);
}

function getLFromRgbColor(r, g, b) {
  return (
    0.2126 * getLFromRgbValue(r) +
    0.7152 * getLFromRgbValue(g) +
    0.0722 * getLFromRgbValue(b)
  );
}

function getLFromHex(hex) {
  var [, r, g, b] = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return getLFromRgbColor(parseInt(r, 16), parseInt(g, 16), parseInt(b, 16));
}

export default function isTextLegibleOverBackground(
  textColor,
  backgroundColor
) {
  var textLightness = getLFromHex(textColor),
    backgroundLightness = getLFromHex(backgroundColor);
  var contrast =
    (Math.max(textLightness, backgroundLightness) + 0.05) /
    (Math.min(textLightness, backgroundLightness) + 0.05);
  return contrast >= 3;
}
