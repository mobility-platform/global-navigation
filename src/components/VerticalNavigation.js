import { setPragma, styled } from "goober";
import { h, Fragment } from "preact";
import { useEffect, useState } from "preact/hooks";
import AppLinks from "./AppLinks";
import Avatar from "./Avatar";
import AvatarItem from "./AvatarItem";
import Brand from "./Brand";
import Burger from "./Burger";
import Links from "./Links";
import { fetchTheme, fetchApplications } from "../utils/api";
import { defaultTheme, ThemeProvider, withTheme } from "./Theme";
import isTextLegibleOverBackground from "../utils/isTextLegibleOverBackground";
import CollapsedLink from "./CollapsedLink";

setPragma(h);

const globalLinks = [
  {
    href: "#",
    label: "My Home",
    icon: `
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="17"
        fill="none"
        viewBox="0 0 16 17"
      >
        <defs />
        <path
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          d="M1 6.3L8 1l7 5.3v8.2c0 .8-.7 1.5-1.6 1.5H2.6c-.9 0-1.6-.7-1.6-1.5V6.2z"
          clip-rule="evenodd"
        />
      </svg>
      `
  },
  {
    href: "#",
    label: "My Organization",
    icon: `
      <svg
        xmlns="http://www.w3.org/2000/svg"
        aria-hidden="true"
        viewBox="0 0 448 512"
      >
        <defs />
        <path d="M352 320a95.6 95.6 0 00-59.8 20.9l-102.5-64a96.6 96.6 0 000-41.7L292.2 171a96 96 0 10-34-54.3l-102.4 64a96 96 0 100 150.2l102.5 64.2A96.3 96.3 0 00256 416a96 96 0 1096-96z" />
      </svg>
      `
  }
];

const Collapsed = withTheme(
  styled("div")(({ theme }) => ({
    position: "fixed",
    top: "0",
    left: "0",
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    width: "60px",
    height: "100%",
    paddingBottom: "20px",
    boxSizing: "border-box",
    boxShadow: "2px 0 8px -3px rgba(0, 0, 0, .2)",
    background: theme.primary,
    color: isTextLegibleOverBackground("#ffffff", theme.primary)
      ? "#ffffff"
      : "#333333",
    fontFamily: `-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol"`
  }))
);

const Extended = withTheme(
  styled("div")(({ theme }) => ({
    position: "fixed",
    top: "0",
    left: "0",
    display: "flex",
    flexDirection: "column",
    width: "100%",
    height: "100%",
    maxWidth: "320px",
    boxShadow: "2px 0 8px -3px rgba(0, 0, 0, .2)",
    transition: "transform 200ms",
    background: theme.background,
    color: isTextLegibleOverBackground("#ffffff", theme.background)
      ? "#ffffff"
      : "#333333",
    fontFamily: `-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol"`
  }))
);

const Shadow = styled("div")(({ isCollapsed }) => ({
  position: "fixed",
  top: "0",
  left: "0",
  width: "100vw",
  height: "100%",
  opacity: isCollapsed ? 0 : 1,
  backgroundColor: "rgba(0, 0, 0, .1)",
  transition: "opacity 200ms",
  pointerEvents: isCollapsed ? "none" : "auto"
}));

const Content = styled("div")(({ isCollapsed }) => ({
  flex: "1",
  display: "flex",
  flexDirection: "column",
  alignItems: isCollapsed ? "center" : "flex-start",
  padding: isCollapsed ? "0" : "0 12px"
}));

const AvatarItemWrapper = styled("div")({
  padding: "8px 16px 8px 16px",
  marginTop: "8px",
  borderTop: "1px solid rgba(0, 0, 0, .1)"
});

const VerticalNavigation = ({ footerLinks, getToken, apiUrl }) => {
  if (!apiUrl || !getToken) {
    throw new Error(
      "`VerticalNavigation` requires the `apiUrl` and `getToken` props. See https://mobility-platform-docs.netlify.com/"
    );
  }

  const [theme, setTheme] = useState(defaultTheme);

  useEffect(() => {
    (async () => {
      const apiTheme = await fetchTheme({ getToken, apiUrl });
      setTheme(apiTheme);
    })();
  }, []);

  const [applications, setApplications] = useState();

  useEffect(() => {
    (async () => {
      const apiApplications = await fetchApplications({ getToken, apiUrl });
      setApplications(apiApplications);
    })();
  }, []);

  const [isCollapsed, setIsCollapsed] = useState(true);
  return (
    <ThemeProvider theme={theme}>
      <Fragment>
        <Shadow
          isCollapsed={isCollapsed}
          onClick={() => setIsCollapsed(true)}
        />

        <Collapsed>
          <Brand isCollapsed={true} />
          <Content isCollapsed={true}>
            <Burger isCollapsed={true} handler={() => setIsCollapsed(false)} />
            <Links isCollapsed={true} data={globalLinks} />
            <AppLinks isCollapsed={true} data={applications} />
            <div
              style={{
                marginTop: "auto",
                display: "flex",
                alignItems: "center",
                flexDirection: "column"
              }}
            >
              <Links isCollapsed={true} data={footerLinks} />
              <CollapsedLink onClick={() => setIsCollapsed(false)}>
                <Avatar src={"https://i.pravatar.cc/40"} size={"22px"} />
              </CollapsedLink>
            </div>
          </Content>
        </Collapsed>

        <Extended
          style={{
            transform: isCollapsed ? "translateX(-100%)" : "translateX(0)"
          }}
        >
          <Brand />
          <div
            style={{
              padding: "10px 16px",
              borderTop: "1px solid rgba(0, 0, 0, .1)",
              borderBottom: "1px solid rgba(0, 0, 0, .1)",
              marginBottom: "16px"
            }}
          >
            <Burger handler={() => setIsCollapsed(true)} />
          </div>
          <Content>
            <Links title={"Global"} data={globalLinks} />
            <div style={{ marginTop: "12px", width: "100%" }}>
              <AppLinks title={"Apps"} data={applications} />
            </div>
            <div style={{ marginTop: "auto", width: "100%" }}>
              <Links title={"Others"} data={footerLinks} />
            </div>
          </Content>
          <AvatarItemWrapper>
            <AvatarItem
              title={"Johanes Does"}
              linkLabel={"Voir le profil"}
              href={"#"}
              target={"_blank"}
            >
              <Avatar src={"https://i.pravatar.cc/40"} size={"40px"} />
            </AvatarItem>
          </AvatarItemWrapper>
        </Extended>
      </Fragment>
    </ThemeProvider>
  );
};

export default VerticalNavigation;
