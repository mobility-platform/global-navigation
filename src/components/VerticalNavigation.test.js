import { render } from "@testing-library/preact";
import { h } from "preact";
import { useFetchConfiguration } from "../utils/Configuration";
import productionConfiguration from "../utils/__fixtures__/production.config.json";
import VerticalNavigation from "./VerticalNavigation";
jest.mock("../utils/Configuration");

describe("VerticalNavigation", () => {
  beforeAll(() => {
    useFetchConfiguration.mockReturnValue(productionConfiguration);
  });

  it("should render without crashing", () => {
    const wrapper = render(
      <VerticalNavigation
        configurationUrl="https://unpkg.com/@mobility-platform/global-navigation-configuration@latest/production.config.json"
        getToken={() => Promise.reject()}
      />
    );
    expect(wrapper.asFragment()).toMatchSnapshot();
    expect(wrapper.asFragment().firstElementChild.className).toBe("mpgn");
  });

  it("should throw an error if configurationUrl or getToken are not passed to the component", () => {
    expect(() => {
      render(<VerticalNavigation />);
    }).toThrow(
      `Global Navigation requires the \`configurationUrl\` and \`getToken\` props. See https://mobility-platform-docs.netlify.com/`
    );
  });
});
