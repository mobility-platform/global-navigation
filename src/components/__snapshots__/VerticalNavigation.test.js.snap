// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalNavigation should render without crashing 1`] = `
<DocumentFragment>
  <div
    class="mpgn"
  />
</DocumentFragment>
`;
